### PR TITLE
Adds database model infrastructure

### PR DESCRIPTION
**Pull request by jacwright**
This contains the basic loading and serialization/deserialization logic. Adds tests. And adds project loading/saving to the projects component.

### CODE REVIEW

File: app/projects/projects.js [hunk 1]
```diff
@@ -1,10 +1,22 @@
 require('./projects.less');
+var Project = require('../../core/project');
```
Comment (matthopson) [inline app/projects/projects.js]: Should we define a root path? May not be that big of an issue - just curious if we'll see a lot of various relative path chains.

File: core/README.md [hunk 2]
```diff
@@ -1,29 +1,34 @@
 # Dabble Core
 
-The core is all the data for a Dabble document. Dabble documents are at a basic level folders and pages. These can be
-very customizable in editing and custom types of folders and pages allow for unique documents such as novels,
-screenplays, research papers, etc.
+The core is all the data for a Dabble project. Dabble projects are at a basic level folders and files. These can be
+very customizable in editing and custom types of folders and files allow for unique types of projects such as novels,
```
Comment (matthopson) [inline core/README.md]: Maybe: ?
"These can be very customizable and allow for unique types of projects..."

File: core/README.md [hunk 3]
```diff
@@ -1,29 +1,34 @@
 # Dabble Core
 
-The core is all the data for a Dabble document. Dabble documents are at a basic level folders and pages. These can be
-very customizable in editing and custom types of folders and pages allow for unique documents such as novels,
-screenplays, research papers, etc.
+The core is all the data for a Dabble project. Dabble projects are at a basic level folders and files. These can be
```
Comment (matthopson) [inline core/README.md]: Also, grammar nit: :stuck_out_tongue_winking_eye: 
"...projects are, at a basic level, folders and files.

File: core/README.md [hunk 4]
```diff
@@ -1,29 +1,34 @@
...
+be a novel, a series, a screenplay, a research project, a white-paper, or more. The project will also contain project
+notes, research, and other items which may not directly be included with the final output of the project.
 
-Folders are a general term that represent any node which contains children. A folder can be a book, a chapter, a folder,
-etc.
+### Folder
 
+Folders are items or nodes in a project's navigation window which contain children. Folders can be generic or represent
+a domain-specific container such as a book or a chapter. Folders have a `type` property that indicates what they
+represent and which indicate how they should be displayed.
```
Comment (matthopson) [inline core/README.md]: "...what they represent `and how` they should be displayed."

File: core/README.md [hunk 5]
```diff
@@ -1,29 +1,34 @@
...
 
-### Page
+### File
 
-A page represents any sub-document that is written on. It can be represented as an actual page or it can be displayed
-as a scene, character sheet, index card, etc.
+A file is any sub-document that contains text. Like folders, files can be generic or they can represent a domain-
+specific item such as a scene, character sheet, or index card. Files have a `type` property that indicates what they
+represent and which indicate how they should be displayed.
```
Comment (matthopson) [inline core/README.md]: Same as previous comment.

File: app/projects/projects.js (hunk 1 above)
Comment (jacwright) [inline app/projects/projects.js]: That's a great idea. Lets add that as another PR.

File: core/db.js [hunk 6]
```diff
@@ -91,8 +96,88 @@ module.exports = function(dbName) {
...
+      var item = saveQueue[id];
+      items.push(item);
+      return item.update;
+    });
+
+    saveQueue = {};
+
+    db.bulkDocs(updates).then(function(bulkResults) {
+      bulkResults.forEach(function(result, i) {
+        var item = items[i];
```
Comment (matthopson) [inline core/db.js]: Do you know for sure you'll always get your results back in the same order?

Comment (jacwright) [inline core/db.js]: It isn't explicit in the PouchDB documentation, but it is in the CouchDB documentation https://wiki.apache.org/couchdb/HTTP_Bulk_Document_API ("The results are returned in the same order as the supplied "docs" array.") which PouchDB copies. And I've found it to be accurate in the tests and past experience.

Comment (jacwright) [inline core/db.js]: And I'm submitting this as a proposed addition to the docs, see https://github.com/pouchdb/pouchdb/pull/4837

Comment (jacwright) [inline core/db.js]: https://github.com/pouchdb/pouchdb/pull/4837#issuecomment-175288603 verifies they are returned in order.

**Pull request merged**